plugins/woocommerce/src/Blocks/Domain/Services/AddressProviderService.php

```diff
@@ -0,0 +1,102 @@
+<?php
+declare( strict_types=1 );
+namespace Automattic\WooCommerce\Blocks\Domain\Services;
+
+use WC_Address_Provider;
+
+/**
+ * Service class for managing address providers.
+ */
+class AddressProviderService {
+
+	/**
+	 * Cached provider class names from the last filter call.
+	 *
+	 * @var string[]
+	 */
+	private $cached_provider_class_names = [];
+
+	/**
+	 * Cached provider instances.
+	 *
+	 * @var WC_Address_Provider[]
+	 */
+	private $cached_providers = [];
+
+	/**
+	 * Get all registered providers.
+	 *
+	 * @return WC_Address_Provider[] array of WC_Address_Providers.
+	 */
+	public function get_registered_providers(): array {
+		/**
+		 * Filter the registered address providers.
+		 *
+		 * @since 10.2.0
+		 * @param WC_Address_Provider[] $providers Array of fully qualified class names that extend WC_Address_Provider.
+		 */
+		$provider_class_names = apply_filters( 'woocommerce_address_providers', [] );
+
+		$logger = wc_get_logger();
+
+		if ( ! is_array( $provider_class_names ) ) {
+			$logger->error(
+				'Invalid return value for woocommerce_address_providers, expected an array of class names.',
+				[
+					'context' => 'address_provider_service',
+				]
+			);
+			return [];
+		}
+
+		// If the class names haven't changed, return the cached instances.
+		if ( $this->cached_provider_class_names === $provider_class_names && ! empty( $this->cached_providers ) ) {
+			return $this->cached_providers;
+		}
+
+		$providers = [];
+
+		foreach ( $provider_class_names as $provider_class_name ) {
+			// Ensure the class exists and is a valid WC_Address_Provider subclass.
+			if ( class_exists( $provider_class_name ) && is_subclass_of( $provider_class_name, WC_Address_Provider::class ) ) {
+				$provider_instance = new $provider_class_name();
+
+				// Validate the instance has the necessary properties.
+				if ( ! empty( $provider_instance->id ) && ! empty( $provider_instance->name ) ) {
```

Comment:
Do we want to ensure IDs are unique? Whats the consequence of providers with duplicate IDs, even if an edge-case? Especially since `is_provider_available()` queries by ID.

Reply:
Good point, we did have that check in the previous iteration, I'll reimplement for the new approach